cornac/models/dnntsp/dnntsp.py

```diff
@@ -0,0 +1,620 @@
+            edge_weight = []
+            for node_1 in nodes.tolist():
+                for node_2 in nodes.tolist():
+                    if (node_1 in basket and node_2 in basket) or (node_1 == node_2):
+                        edge_weight.append(edges_weight_dict.get((node_1, node_2), 0.0))
+                    else:
+                        edge_weight.append(0.0)
+            edges_weight.append(torch.Tensor(edge_weight))
+        batch_edges_weight.append(torch.stack(edges_weight).to(device))
+    return (
+        batch_g,
+        batch_nodes_feature,
+        batch_edges_weight,
+        batch_lengths,
+        batch_nodes,
+        batch_targets,
+    )
+
+
+class BPRLoss(nn.Module):
+    def __init__(self):
+        super(BPRLoss, self).__init__()
+
+    def forward(self, predict, truth):
+        """
+        Args:
+            predict: (batch_size, items_total) / (batch_size, baskets_num, item_total)
+            truth: (batch_size, items_total) / (batch_size, baskets_num, item_total)
+        Returns:
+            output: tensor
+        """
+        result = self.batch_bpr_loss(predict, truth)
+
+        return result
+
+    def batch_bpr_loss(self, predict, truth):
+        """
+        Args:
+            predict: (batch_size, items_total)
+            truth: (batch_size, items_total)
+        Returns:
+            output: tensor
+        """
+        items_total = truth.shape[1]
+        nll = 0
+        for user, predictUser in zip(truth, predict):
+            pos_idx = user.clone().detach()
+            preUser = predictUser[pos_idx]
+            non_zero_list = list(itertools.chain.from_iterable(torch.nonzero(user)))
+            random_list = list(set(range(0, items_total)) - set(non_zero_list))
+            random.shuffle(random_list)
+            neg_idx = torch.tensor(random_list[: len(preUser)])
+            score = preUser - predictUser[neg_idx]
+            nll += -torch.mean(torch.nn.LogSigmoid()(score))
+        return nll
+
+
+class WeightMSELoss(nn.Module):
+    def __init__(self, weights=None):
+        """
+        Args:
+            weights: tensor, (items_total, )
+        """
+        super(WeightMSELoss, self).__init__()
+        self.weights = weights
+        if weights is not None:
+            self.weights = torch.sqrt(weights)
+        self.mse_loss = nn.MSELoss(reduction="sum")
+
+    def forward(self, predict, truth):
+        """
+        Args:
+            predict: tenor, (batch_size, items_total)
+            truth: tensor, (batch_size, items_total)
+        Returns:
+            output: tensor
+        """
+        predict = torch.sigmoid(predict)
+        truth = truth.float()
+        if self.weights is not None:
+            self.weights = self.weights.to(truth.device)
+            predict = predict * self.weights
+            truth = truth * self.weights
+
+        loss = self.mse_loss(predict, truth)
+        return loss
+
+
+class W_multilabel(nn.Module):
```

Comment:
Is this redundant?

Reply:
Yes.
I have removed the class in the last commit.